klippy/extras/bed_mesh.py

```diff
@@ -415,6 +444,80 @@ def print_probed_positions(self, print_func):
+
+
+    def tilt_probe_finalize(self, offsets, positions):
+        x_offset = offsets[0]
+        y_offset = offsets[1]
+        z_offset = offsets[2]
+        self.gcode.respond_info \
+                ("offsets: %f %f %f" % (x_offset,y_offset,z_offset) );
+
+        params=self.bedmesh.z_mesh.probe_params
+
+        if self.relative_reference_index is not None:
+            # zero out probe z offset and
+            # set offset relative to reference index
+            z_offset = positions[self.relative_reference_index][2]
+
+        if self.probed_z_table_backup is None:
+            self.probed_z_table_backup = copy.deepcopy(self.probed_z_table)
+
+        t_probed_z_table=copy.deepcopy(self.probed_z_table_backup)
+        self.bedmesh.z_mesh.build_mesh(t_probed_z_table)
+        pts=[]
+        # shift measured positions
+        for pos in positions:
+            # offset according to the mesh
+            calc_Zval=self.bedmesh.z_mesh.\
+                    calc_z(pos[0]+x_offset,pos[1]+y_offset)+z_offset
+            ZvalCorrectioncorr=pos[2] -calc_Zval;
+            pts.append([pos[0],pos[1],ZvalCorrectioncorr] )
+
+        # cross product of vectors defined by 3 probed points
+        cx=pts[1][2]*(pts[0][1] - pts[2][1]) + \
+                pts[0][2]*(-pts[1][1] + pts[2][1]) + \
+                (-pts[0][1] + pts[1][1])*pts[2][2]
+        cy=pts[0][2]*(pts[1][0] - pts[2][0]) +\
+                pts[1][2]*(-pts[0][0] + pts[2][0]) + \
+                ( pts[0][0] - pts[1][0])*pts[2][2]
+        cz=pts[1][1]*(pts[0][0] - pts[2][0]) + \
+                pts[0][1]*(-pts[1][0] + pts[2][0]) + \
+                (-pts[0][0] + pts[1][0])*pts[2][1]
+        # self.gcode.respond_info("Normal to plane is [%f,%f,%f]" % (cx,cy,cz))
+        # calcuate d in the equation of plane cx*X+cy*Y+cz*Zd==d
+        d=cx*pts[0][0]+cy*pts[0][1]+cz*pts[0][2]
+
+        # now for any (x,y) z defines as (d-cx*X-cy*Y)/cz
+        z_correction=[-cx/cz,-cy/cz,d/cz]
+
+        x_cnt = params['x_count']
+        y_cnt = params['y_count']
+        min_x=params['min_x']
+        min_y=params['min_y']
+        x_dist = (params['max_x'] -params['min_x']) / (x_cnt - 1)
+        y_dist = (params['max_y'] -params['min_y']) / (y_cnt - 1)
+        for i in range(x_cnt):
+            for j in range(y_cnt):
+                xx=(min_x+i*x_dist)
+                yy=(min_y+j*y_dist)
+                Zcorr=z_correction[0]*xx+z_correction[1]*yy+z_correction[2]
+                t_probed_z_table[j][i]+=Zcorr
+
+        mesh = ZMesh(params)
+        try:
+            mesh.build_mesh(t_probed_z_table)
+        except BedMeshError as e:
+            self.probed_z_table = copy.deepcopy(self.probed_z_table_backup)
+            raise self.gcode.error(e.message)
+        self.bedmesh.set_mesh(mesh)
+
+        self.probed_z_table=t_probed_z_table
+
+        self.gcode.respond_info("Mesh Bed Tilting Complete")
+        self.save_profile("default")
```

Comment:
I don't think you should save the profile here.  It seems to me that BED_MESH_TILT is an operation that should be applied to an already saved mesh which is a reference that should not be changed.  If the user wishes then they can save the profile manually.

Reply:
This is something where you definitely can tell better. Initially I didn't save. But then I got impression that it doesn't save to config, it just saves to session internals. If resulting mesh is going to be applied anyway, then I agree that "self.save_profile("default")" should be removed.  Keep in mind, I don't know klipper internals, and I even don't really know python :) I just put here math with reference implementation. As you could see I make backup of probed table, so I agree that transformation has to be applied to original probed mesh.